akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala

```diff
@@ -0,0 +1,206 @@
+
+        partnerRef match {
+          case OptionVal.Some(ref) ⇒
+            ref ! StreamRefsProtocol.OnSubscribeHandshake(self.ref)
+            tryPull()
+          case _ ⇒ // nothing to do
+        }
+
+        scheduleOnce(SubscriptionTimeoutTimerKey, subscriptionTimeout.timeout)
+      }
+
+      lazy val initialReceive: ((ActorRef, Any)) ⇒ Unit = {
+        case (_, Terminated(ref)) ⇒
+          if (ref == getPartnerRef)
+            failStage(RemoteStreamRefActorTerminatedException(s"Remote target receiver of data $partnerRef terminated. " +
+              s"Local stream terminating, message loss (on remote side) may have happened."))
+
+        case (sender, StreamRefsProtocol.CumulativeDemand(d)) ⇒
+          observeAndValidateSender(sender, "Illegal sender for CumulativeDemand")
+
+          if (remoteCumulativeDemandReceived < d) {
+            remoteCumulativeDemandReceived = d
+            log.debug("Received cumulative demand [{}], consumable demand: [{}]", StreamRefsProtocol.CumulativeDemand(d), remoteCumulativeDemandReceived - remoteCumulativeDemandConsumed)
+          }
+
+          tryPull()
+      }
+
+      override def onPush(): Unit = {
+        val elem = grabSequenced(in)
+        getPartnerRef ! elem
+        log.debug("Sending sequenced: {} to {}", elem, getPartnerRef)
+        tryPull()
+      }
+
+      private def tryPull() =
+        if (remoteCumulativeDemandConsumed < remoteCumulativeDemandReceived && !hasBeenPulled(in)) {
+          pull(in)
+        }
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case SubscriptionTimeoutTimerKey ⇒
+          val ex = StreamRefSubscriptionTimeoutException(
+            // we know the future has been competed by now, since it is in preStart
+            s"[$stageActorName] Remote side did not subscribe (materialize) handed out Sink reference [${promise.future.value}], " +
+              s"within subscription timeout: ${PrettyDuration.format(subscriptionTimeout.timeout)}!")
+
+          throw ex // this will also log the exception, unlike failStage; this should fail rarely, but would be good to have it "loud"
+      }
+
+      private def grabSequenced[T](in: Inlet[T]): StreamRefsProtocol.SequencedOnNext[T] = {
+        val onNext = StreamRefsProtocol.SequencedOnNext(remoteCumulativeDemandConsumed, grab(in))
+        remoteCumulativeDemandConsumed += 1
+        onNext
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit =
+        partnerRef match {
+          case OptionVal.Some(ref) ⇒
+            ref ! StreamRefsProtocol.RemoteStreamFailure(ex.getMessage)
+            self.unwatch(getPartnerRef)
+            super.onUpstreamFailure(ex)
+
+          case _ ⇒
+            completedBeforeRemoteConnected = OptionVal(scala.util.Failure(ex))
+            // not terminating on purpose, since other side may subscribe still and then we want to fail it
+            setKeepGoing(true)
```

Comment:
How is it ever stopped then? Couldn't the other side not just watch the ref instead?

Reply:
Hmm, I see, either by timeout or by the below handling in `observeAndValidateSender`. Maybe add a comment to those setKeepGoings then to make that clear.

Reply:
added